Kernel/KSyms.cpp

```diff
@@ -147,7 +153,13 @@ NEVER_INLINE static void dump_backtrace_impl(FlatPtr base_pointer, bool use_ksym
         void* fault_at;
         FlatPtr copied_stack_ptr[2];
         FlatPtr* stack_ptr = (FlatPtr*)base_pointer;
+#if ARCH(X86_64) || ARCH(AARCH64)
         while (stack_ptr && safe_memcpy(copied_stack_ptr, stack_ptr, sizeof(copied_stack_ptr), fault_at)) {
+#elif ARCH(RISCV64)
+        while (stack_ptr && safe_memcpy(copied_stack_ptr, stack_ptr - 2, sizeof(copied_stack_ptr), fault_at)) {
```

Comment:
can we give the magic value 2 a name, like `RISCV_FRAME_POINTER_OFFSET` (the caps hurt my eyes ) or something? otherwise this seems reasonable.

Reply:
or... perhaps we should calculate `base_pointer` differently?

Reply:
I am not sure how to best refactor this magic number out would be. We already use magic indices on the copied_stack_ptr array.

The -2 is needed because on RISC-V `fp` points to the old `sp`, `fp-1*8` is the return addr and `fp-2*8` is the old `fp`.
so essentially this copied_stack_ptr array is shifted by 2*8 Bytes.

This is kind of a hack though, but otherwise we would have to probably make that function arch-specific.

Reply:
I could also refactor that code to not use magic indices on that array and rather copy the two values from the stack into retaddr and stack_ptr immediatly after the memcpy.

although stack_ptr is really the frame_ptr, don't know what x86 would call that here.

Reply:
Are these comments good enough?
I don't see any nice way to refactor that code though.

Reply:
The alternative would be to do something like this:
```cpp
#if ARCH(X86_64) || ARCH(AARCH64)
    FlatPtr previous_frame_pointer_and_return_address[2];
    <copy that array from stack_ptr>
    stack_ptr = previous_frame_pointer_and_return_address[0];
    retaddr = previous_frame_pointer_and_return_address[1];
#elif ARCH(RISCV64)
    FlatPtr previous_frame_pointer_and_return_address[2];
    <copy that array from stack_ptr-2>
    stack_ptr = previous_frame_pointer_and_return_address[0];
    retaddr = previous_frame_pointer_and_return_address[1];
#endif
```
That would only introduce unnecessary code duplication

Reply:
What do you think about this now?

Reply:
ah, meant to comment. yeah the comments look good enough. maybe we can make it prettier in the future.

Reply:
I refactored it again as I wrote that comment.
Did you notice that? I don't want you to merge it by accident.

Reply:
nope, I didn't notice it, thanks! I looked over it again and I do like the FrameRecord idea. it makes it clear what values we are grabbing and from where 👍